### PR TITLE
[corlib] CoreFX sources for System namespace

### PR DESCRIPTION
**Pull request by EgorBo**
Brings CoreFX and CoreRT sources for primitive types inside corlib (and other types under System namespace)

*ADS:* I use this tool to inspect all files I replace https://github.com/EgorBo/MonoMergeHelper 🙂 

Part of #7737

### CODE REVIEW

File: mcs/class/corlib/corlib.dll.sources [hunk 1]
```diff
@@ -91,8 +91,8 @@ System/AppDomainSetup.cs
 System/ApplicationIdentity.cs
 System/ArgIterator.cs
 System/Array.cs
-System/AssemblyLoadEventArgs.cs
-System/AssemblyLoadEventHandler.cs
+../../../external/corefx/src/Common/src/CoreLib/System/AssemblyLoadEventArgs.cs
```
Comment (luhenry) [inline mcs/class/corlib/corlib.dll.sources]: Keeping these `../../../external/corefx` paths mixed with the "local" and the referencesource one looks messy when you look at the `.dll.sources` file. It's already been fixed previously https://github.com/mono/mono/pull/8389. I don't want you to require to change it since it's not really breaking anything, but it would be better over the long run to move these files into their own block with the rest of the `../../../external/corefx` paths.

Comment (EgorBo) [reply mcs/class/corlib/corlib.dll.sources]: @luhenry ok, I'll re-order them before merge, I am replacing sources via tool I mentioned in the description

**Comment by EgorBo** [comment]
@monojenkins build Linux x64

**Comment by marek-safar** [comment]
There is still a lot of extra TypeForwardedFromAttribute, could you also add tests for all non trivial types we are switching

**Comment by EgorBo** [comment]
@marek-safar yeah I know, the today's mono/corefx patch was for a different PR.
I'll file one more tomorrow to hide these ones.

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
btw, there is still one more serialization forwarded at InvalidComObjectException

**Comment by EgorBo** [comment]
So it looks like currently there are two issues left:
1) WebAssembly - not sure what the error means
2) Decimal.cs doesn't compile when `mcs` is used

**Comment by kumpera** [comment]
@EgorBo it means the test suite aborted at some point - aborts are non-fatal on wasm.

Here's the error:

```
WASM: CANNOT HANDLE COOKIE IDIII
WASM: * Assertion at /mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/mini/wasm_m2n_invoke.g.h:535, condition `0' not met
```

It's a trivial fix, I'll push a PR with a fix tomorrow.

**Comment by kumpera** [comment]
@EgorBo https://github.com/mono/mono/pull/9107

File: mcs/class/referencesource/mscorlib/system/convert.cs [hunk 2]
```diff
@@ -2472,7 +2476,7 @@ public static decimal ToDecimal(DateTime value)
         [SecurityCritical]
         private static unsafe Int32 FromBase64_Decode(Char* startInputPtr, Int32 inputLength, Byte* startDestPtr, Int32 destLength) {
 
-            // You may find this method weird to look at. It�s written for performance, not aesthetics.
+            // You may find this method weird to look at. It’s written for performance, not aesthetics.
```
Comment (akoeplinger) [inline mcs/class/referencesource/mscorlib/system/convert.cs]: why this change?

File: mcs/class/corlib/corert/Interop.Number.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+using System.Runtime.CompilerServices;
+
+internal static partial class Interop
+{
+    internal unsafe partial class Sys
+    {   
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal static extern unsafe int DoubleToString(double value, byte* format, byte* buffer, int bufferLength);
+    }
+}
```
Comment (akoeplinger) [inline mcs/class/corlib/corert/Interop.Number.cs]: instead of copying the file we should make #if MONO in the corert version (like we did with Interop.RunLoop.cs).

File: mcs/class/corlib/corert/RuntimeImports.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime
+{
+    public static class RuntimeImports
+    {
+        [DllImport("__Internal", ExactSpelling = true)]
+        internal static extern unsafe void _ecvt_s(byte* buffer, int sizeInBytes, double value, int count, int* dec, int* sign);
```
Comment (akoeplinger) [inline mcs/class/corlib/corert/RuntimeImports.cs]: what does this end up calling to?

File: mcs/class/corlib/Test/System/NumberFormatterTest.cs [hunk 5]
```diff
@@ -1567,6 +1568,7 @@ public void Test05028 ()
 		}
 
 		[Test]
+		[Category("NotWorking")] // bug in .NET Core: https://github.com/dotnet/coreclr/issues/18218
```
Comment (akoeplinger) [inline mcs/class/corlib/Test/System/NumberFormatterTest.cs]: seems like a fix was already merged: https://github.com/dotnet/coreclr/pull/18221, can we cherrypick this into our corefx fork and reenable the test?

File: mcs/class/referencesource/mscorlib/system/convert.cs (hunk 2 above)
Comment (EgorBo) [reply mcs/class/referencesource/mscorlib/system/convert.cs]: I did not change it - seems like IDE did it (vscode or rider) 🙂

File: mcs/class/corlib/corert/RuntimeImports.cs (hunk 4 above)
Comment (EgorBo) [reply mcs/class/corlib/corert/RuntimeImports.cs]: @akoeplinger https://msdn.microsoft.com/en-us/library/ehz2dset.aspx (c run-time library) - it's needed only for Windows - https://github.com/dotnet/corert/blob/e79755bebe745c1229df8e5ba9a20b4fa3fc950d/src/System.Private.CoreLib/src/System/Number.Windows.cs#L27 and https://github.com/dotnet/corert/blob/df8d48011ebc11c5c97b8384275d40ef4f95388e/src/System.Private.CoreLib/src/System/Runtime/RuntimeImports.cs#L1015 but I am not quite sure why we can't just use `snprintf` icall on all platforms

File: mcs/class/corlib/corert/Interop.Number.cs (hunk 3 above)
Comment (EgorBo) [reply mcs/class/corlib/corert/Interop.Number.cs]: yeah probably makes sense

File: mcs/class/corlib/Test/System/NumberFormatterTest.cs (hunk 5 above)
Comment (EgorBo) [reply mcs/class/corlib/Test/System/NumberFormatterTest.cs]: Sure!

File: mono/metadata/icall-def.h [hunk 6]
```diff
@@ -789,6 +792,9 @@ ICALL(REMSER_1, "InternalExecute", ves_icall_InternalExecute)
 HANDLES(ICALL(REMSER_2, "IsTransparentProxy", ves_icall_IsTransparentProxy))
 #endif
 
+ICALL_TYPE(CLR_RUNTIME_IMPORTS, "System.Runtime.RuntimeImports", CLR_RUNTIME_IMPORTS_1)
+ICALL(CLR_RUNTIME_IMPORTS_1, "_ecvt_s", ves_icall_Mono_RuntimeImports_ecvt_s)
```
Comment (luhenry) [inline mono/metadata/icall-def.h]: `ves_icall_System_Runtime_RuntimeImports_ecvt_s`

Comment (EgorBo) [reply mono/metadata/icall-def.h]: yeah, copy-paste :(

**Comment by EgorBo** [comment]
@monojenkins build failed

File: acceptance-tests/coreclr.mk [hunk 7]
```diff
@@ -3242,7 +3241,6 @@ CORECLR_COREMANGLIB_TEST_CS_SRC=		\
 	$(CORECLR_PATH)/tests/src/CoreMangLib/cti/system/security/securityexception/securityexceptionctor3.cs	\
 	$(CORECLR_PATH)/tests/src/CoreMangLib/cti/system/security/securityexception/securityexceptiontostring.cs	\
 	$(CORECLR_PATH)/tests/src/CoreMangLib/cti/system/single/singleepsilon.cs	\
-	$(CORECLR_PATH)/tests/src/CoreMangLib/cti/system/single/singlegethashcode.cs	\
```
Comment (EgorBo) [inline acceptance-tests/coreclr.mk]: see https://github.com/dotnet/coreclr/pull/16567

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
A small effort to test whether CoreRT's Decimal works for CoreCLR or not: https://github.com/dotnet/coreclr/pull/18868

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@EgorBo please rebase and bump monolite version

**Comment by marek-safar** [comment]
@monojenkins build deb with monolite

**Comment by akoeplinger** [comment]
@monojenkins commit apidiff and csproj
